keiko-sql/src/main/kotlin/com/netflix/spinnaker/q/sql/SqlQueue.kt

```diff
@@ -402,24 +403,46 @@ class SqlQueue(
       jooq.transaction { config ->
         val txn = DSL.using(config)
 
-        txn.insertInto(messagesTable)
-          .set(idField, ulid.toString())
-          .set(fingerprintField, fingerprint)
-          .set(bodyField, mapper.writeValueAsString(message))
-          .set(updatedAtField, clock.millis())
-          .onDuplicateKeyUpdate()
-          .set(idField, MySQLDSL.values(idField) as Any)
-          .set(bodyField, MySQLDSL.values(bodyField) as Any)
-          .execute()
+        val existingFingerprint = txn.select(field(idField))
+          .from(unackedTable)
+          .where(fingerprintField.eq(fingerprint))
+          .fetchOne(0, String::class.java)
+
+        // If the fingerprint exists in the unacked table, then the message already exists. In this
+        // case, we'll reschedule the message for the delay amount, rather than adding a new
+        // message onto the queue.
+        if (existingFingerprint != null) {
+          // This is a small optimization: If the delay is less than 1 second, do nothing.
+          if (delay.get(ChronoUnit.SECONDS) < 1L) {
```

Comment:
Not sure if this is right, a delay of 0 means queue for immediate execution. Since we aren't looking at the delivery value of a possible row on the db, it could be much further out, so we'd still want to update it if delay == 0.